sig-azure/0018-20180711-azure-availability-zones.md

```diff
@@ -0,0 +1,238 @@
+
+The node's zone could get by ARM API or instance metadata. This will be added in  `GetZoneByProviderID()` and `GetZoneByNodeName()`.
+
+### Get by instance metadata
+
+This method is used in kube-controller-manager.
+
+```sh
+# Instance metadata API should be upgraded to 2017-12-01.
+$ curl -H Metadata:true "http://169.254.169.254/metadata/instance/compute/zone?api-version=2017-12-01&format=text"
+2
+```
+
+### Get by Go SDK
+
+This method is used in cloud-controller-manager.
+
+No `zones` property is included in `VirtualMachineScaleSetVM` yet in Azure Go SDK (including latest 2018-04-01 compute API).
+
+We need to ask Azure Go SDK to add `zones` for `VirtualMachineScaleSetVM`. Opened the issue https://github.com/Azure/azure-sdk-for-go/issues/2183 for tracking it.
+
+> Note: there's already `zones` property in `VirtualMachineScaleSet`, `VirtualMachine` and `Disk`.
+
+## LoadBalancer and PublicIP
+
+LoadBalancer with standard SKU will be created and all available nodes (including VirtualMachines and VirtualMachineScaleSetVms, together with optional masters configured via excludeMasterFromStandardLB) are added to LoadBalancer backend pools.
+
+PublicIPs will also be created with standard SKU, and they are zone redundant by default.
+
+Note that zonal PublicIPs are not supported. We may add this easily if there’re clear use-cases in the future.
+
+## AzureDisk
+
+When Azure managed disks are created, the `PersistentVolumeLabel` admission controller automatically adds zone labels to them. The scheduler (via `VolumeZonePredicate`) will then ensure that pods that claim a given volume are only placed into the same zone as that volume, as volumes cannot be attached across zones.
+
+> Note that only managed disks are supported. Blob disks don't support availability zones on Azure.
+
+### PVLabeler
+
+`PVLabeler` interface should be implemented for AzureDisk:
+
+```go
+// PVLabeler is an abstract, pluggable interface for fetching labels for volumes
+type PVLabeler interface {
+    GetLabelsForVolume(ctx context.Context, pv *v1.PersistentVolume) (map[string]string, error)
+}
+```
+
+It should return the region and zone of the AzureDisk, e.g.
+
+- `failure-domain.beta.kubernetes.io/region=centralus`
+- `failure-domain.beta.kubernetes.io/zone=centralus-1`
+
+so that the PV will be created with labels:
+
+```sh
+$ kubectl get pv --show-labels
+NAME           CAPACITY   ACCESSMODES   STATUS    CLAIM            REASON    AGE       LABELS
+pv-managed-abc 5Gi        RWO           Bound     default/claim1             46s       failure-domain.beta.kubernetes.io/region=centralus,failure-domain.beta.kubernetes.io/zone=centralus-1
+```
+
+### PersistentVolumeLabel
+
+Besides PVLabeler interface, [PersistentVolumeLabel](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/persistentvolume/label/admission.go) admission controller should also updated with AzureDisk support, so that new PVs could be applied with above labels automatically.
```

Comment:
This admission controller is deprecated now and I believe its replacement is in the cloud controller manager

Reply:
Sorry, didn't notice this has been deprecated. Cloud controller manager will GA in v1.13 or v1.4. What's the replacement before its GA?

Reply:
I'm not sure if there is a replacement in the meantime, but both methods should be implemented

Reply:
@msau42 you're right, 

- if cloud-controller-manager is used, then that is actually PVLabeler interface which is used persistent volume label controller
- or else, this admission controller is used.

Will add this info to both this and PVLabler part.